src/main/java/com/google/sps/servlets/CommentService.java

```diff
@@ -0,0 +1,77 @@
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.services.youtube.model.CommentThread;
+import com.google.api.services.youtube.model.CommentThreadListResponse;
+import com.google.api.services.youtube.YouTube;
+import java.io.IOException;
+import java.lang.Exception;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Collections;
+
+public class CommentService {
+
+  private final String DEVELOPER_KEY = "DEV_KEY";
+  private static final String APPLICATION_NAME = "YouTube Comments";
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+  private YouTube getService() throws GeneralSecurityException, IOException{
+    final NetHttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    return new YouTube.Builder(httpTransport, JSON_FACTORY, null)
+        .setApplicationName(APPLICATION_NAME)
+        .build();
+  }
+
+  /**
+   * Calls YouTube Data API to return a list of the top 25 comments for a video, identified by its
+   * video id. 
+   */
+  public List<String> getCommentsFromId(String videoId)
+      throws GeneralSecurityException, IOException, GoogleJsonResponseException {
+    YouTube youtubeService = getService();
+    YouTube.CommentThreads.List request = youtubeService.commentThreads()
+        .list(Collections.singletonList("snippet"));
+    try { 
+      // Call YouTube API and get the 25 comments that appear first in comment section of the video
+      CommentThreadListResponse response = request.setKey(DEVELOPER_KEY)
+          .setMaxResults(25L)
+          .setOrder("relevance")
+          .setTextFormat("plainText")
+          .setVideoId(videoId)
+          .execute();
+      
+      if (response.getItems().isEmpty()) {
+        throw new Exception(String.format("Video %s has no comments.", videoId));
+      }
+
+      List<String> commentsList = new ArrayList<>();
+      for (CommentThread commentThread : response.getItems()){
+        String commentContent = commentThread.getSnippet()
+            .getTopLevelComment()
+            .getSnippet()
+            .getTextDisplay();
+        commentsList.add(commentContent);
+      }
+  
+      return commentsList;
+
+    } catch (GoogleJsonResponseException e) {
+      System.out.println(e.getDetails().getMessage());
+      // If comments cannot be retrieved, return null to indicate that sentiment score should not
+      // be calculated. An arbitrary score of 0 will lower the average sentiment score between
+      // captions and comments scores. 
+      return null;
+    }
+    catch (Exception e) {
+      System.out.println(e.getMessage());
+      return null;
+    }
```

Comment:
Generic Exception catches aren't recommended. Do you have a list of all the potential exceptions that this block can throw?  Do some of them need to be handled differently? Are some of them avoidable?

You can also specify catch blocks to catch any exceptions from a given list.

Reply:
I think GoogleJsonResponseException is thrown for all the errors that can happen, so we'll return an empty list in those cases. I decided to take out the generic exception

For cases where the video id isn't valid or the video is private/can't be accessed, we'll be checking for those cases in the front end, so do I still handle that case in the backend as well?

Reply:
Can't assume any checks work in the front end. Anyone can send a POST GET request directly here without going via the front end code. I recommend you do a validation check in this servlet as well. You shouldn't make it yourself though. Either call another YouTube API service that validates it or if you the call you already make in this PR tells you that it is an invalid id, you should catch it and throw an error upwards saying it was an invalid argument.